### PR TITLE
Add channel facility for inter-process communication

### PR DESCRIPTION
**Pull request by wingo**
Change the lwAFTR to use channels instead of signals, timers, and
syscalls.  This way we avoid kernel jitter.

### CODE REVIEW

File: src/apps/lwaftr/lwaftr.lua [hunk 1]
```diff
@@ -599,6 +581,21 @@ function LwAftr:push ()
    local o4, o6 = self.output.v4, self.output.v6
    self.o4, self.o6 = o4, o6
 
+   do
+      local msg = self.control:pop()
+      if msg then
+         if msg == messages.lwaftr_message_reload then
```
Comment (wingo) [inline src/apps/lwaftr/lwaftr.lua]: so, this is bogus of course -- we need to be checking msg.kind == messages.lwaftr_message_reload

**Comment by kbara** [comment]
I like the idea behind this patch. I have some qualms about the practical side of things.
a) It's one day before a release. (subpoint: and CI fails with this patch).
b) How expensive is a full memory barrier? How confident are you that you are using exactly the amount of them necessary, in the right places? Why?
c) Do we really want to get rid of the classical unix SIGHUP mechanism and replace it with a ```snabb-lwaftr control``` mechanism? I agree with you that the syscall overhead is annoying, but is it acceptable, say, once per second, given that we're using it for binding table updates?
d) The comments include:
```
+--  2. It should be possible for the reader to update the read pointer
+--     without stompling other memory, notably the write pointer.
+--  3. It should be possible for the writer to update the write pointer
+--     without stompling other memory, notably the read pointer.
+-- We throw in a full memory barrier in a hope that it ensures (1) and
+-- (4).  I do not know about (2) and (3).
```
I think you have to know before this is mergable. (For a one-reader-one-writer scenario, I believe that this particular point is fine, but having a clearly documented mental model is non-optional).

----
I like channels. I like this code. It needs more work.

**Comment by wingo** [comment]
Very good concerns!  Thanks for the review :)

(a) Yep!  Need to be careful about this one.
(b) I have measured throughput and the barriers seem to have no effect on the `snabb-lwaftr bench` workload relative to master.  However compared to `lwaftr_starfruit` where we are doing the read() 200 times a second the latency variability is quite a bit improved.  Here is an instrumented measurement of min/max/average jitter for `lwaftr_starfruit` on an isolated CPU on our snabb1 test machine: 

```
89.658002us min / 170.273001us max / 91.179489us avg
89.511006us min / 113.435002us max / 91.138392us avg
89.619003us min / 193.275999us max / 91.152250us avg
89.573005us min / 111.829002us max / 91.149638us avg
89.609995us min / 225.863005us max / 91.221252us avg
89.687004us min / 113.907001us max / 91.259319us avg
89.648995us min / 113.179005us max / 91.233124us avg
89.585999us min / 115.279996us max / 91.193671us avg
89.638001us min / 111.398003us max / 91.262545us avg
89.667999us min / 211.437997us max / 91.258542us avg
89.647998us min / 142.939003us max / 91.157984us avg
89.556001us min / 110.507994us max / 91.086688us avg
89.598994us min / 106.795997us max / 91.109673us avg
89.592999us min / 279.783002us max / 91.163919us avg
89.585999us min / 110.869994us max / 91.088377us avg
89.554997us min / 104.858002us max / 91.115229us avg
89.581001us min / 107.526001us max / 91.128466us avg
89.563997us min / 104.646002us max / 91.068837us avg
89.542998us min / 109.241999us max / 91.068947us avg
89.556001us min / 121.158999us max / 91.124059us avg
89.497997us min / 116.511001us max / 91.035719us avg
89.577996us min / 127.334999us max / 91.134305us avg
89.477995us min / 107.960004us max / 91.144158us avg
89.560002us min / 106.364998us max / 91.113032us avg
89.467998us min / 104.280000us max / 91.101229us avg
89.508998us min / 115.686002us max / 91.117078us avg
```

These printouts are every 10000 breaths, or every second or so.  (Would be nice to get a metrics patch like this upstream; some other time :)

For this branch:

```
78.949000us min / 95.834999us max / 80.110727us avg
78.983998us min / 94.360999us max / 80.108170us avg
78.962999us min / 95.991003us max / 80.099972us avg
78.972000us min / 101.450998us max / 80.097394us avg
78.919998us min / 96.060998us max / 80.090966us avg
78.970996us min / 95.030002us max / 80.099854us avg
78.933001us min / 95.314994us max / 80.155845us avg
79.047997us min / 94.124000us max / 80.167464us avg
78.990997us min / 93.533003us max / 80.121075us avg
78.978999us min / 99.666002us max / 80.101706us avg
78.971003us min / 94.742005us max / 80.090174us avg
78.953999us min / 97.185999us max / 80.139818us avg
78.997000us min / 94.101000us max / 80.157400us avg
79.041994us min / 98.688004us max / 80.122312us avg
78.968005us min / 94.283001us max / 80.119475us avg
78.967998us min / 101.110003us max / 80.107742us avg
78.949997us min / 93.556002us max / 80.098038us avg
78.992998us min / 97.285003us max / 80.107042us avg
78.949000us min / 95.834999us max / 80.110727us avg
```

I think will go a long way towards solving our jitter issues while still remaining responsive to external commands.

(c) Here we veer into opinion territory, but in my opinion yes we want to get rid of signals.  Signals are a terrible interface: they convey no additional information other than "hey here's a wee integer".  You can't tell a running lwaftr what file to dump a config to, what binding table to reload, nothing.  Given that they are insufficient for the kinds of interactions we'd like to have with the lwaftr, and unnecessary given this facility, I would certainly remove any kind of support for them from the lwaftr, yes.

(d) I read up a bit and I think I have the bases covered now.

---

In summary I guess I feel pretty good about this patch.  I don't expect problems from it, and I feel comfortable releasing with it.

**Comment by kbara** [comment]
If you're also convinced about points 2 and 3, with regard to not ```stompling`` memory, we can give this a go. But this is really the LAST release that I'm going to ack anything that isn't a critical bug fix after the code freeze date. The farther this gets from being alpha software, the more that kind of thing is important to actually stick to.

**Comment by kbara** [comment]
Also: nice work on reducing jitter. :-)

**Comment by lukego** [comment]
JFYI: I spent most of last easter tracking down a bug that turned out to be a missing memory barrier. I think that writes from multiple cores crossing "in flight" is the tricky case on x86. http://dpdk.org/ml/archives/dev/2015-April/016219.html

**Comment by wingo** [comment]
@lukego If you get a chance to look at  channel.lua, any thoughts would be very welcome.  There's a big ol' comment that explains what's going on.  Lines 93 to 160.

**Comment by kbara** [comment]
@lukego yep, that kind of correctness issue is my biggest fear on this. Writes aren't re-orderable in a way that can hurt here, and nothing stood out as clearly wrong, but ....

**Comment by lukego** [comment]
bit tired but some thoughts:

In this example I think that compiler barriers are needed but not hardware barriers (on x86). The important thing is that the data is stored before the index is updated so that the other side does not access the payload before it is written. x86 does not reorder writes and would not do that but LuaJIT might.

I suspect you need to move the barriers to be after the payload is written but before the index is updated i.e. lexically up one line in the source file. That would ensure that the other core does not see the index update before the payload. Now the barrier is after both and so it seems like LuaJIT could decide to flush the store to the index before you want it to.

If you wanted to run this at really high speeds then you might want to worry about "false sharing" i.e. two cores writing to the same cache line. Resolution would be to move the read and write state onto separate cache lines (align to 64 bytes).

Finally: it seems like you could also use a vanilla interprocess `struct link` for this use case in the future. It's basically the same data structure, right? Messages could be packets. cc @xrme.

**Comment by lukego** [comment]
another thought: this is basically a "kernel bypass" implementation of a pipe or a unix socket, right?

I wonder: can we really not use kernel mechanisms for this? why not? I understand that you have significantly reduced latency with this somehow but I don't know the details.

My current mental model is that a system call is expensive (~100 cycles) but not that it fundamentally introduces latency. If we really want to avoid *any* system calls then that would have pretty major implications. I would want to look really hard for a simpler alternative first e.g. to `nice` the process to realtime priority, etc.

Could take this discussion if/when you upstream this code and people can ask: why not use a non-blocking unix socket?
